### ISSUE TITLE
Consider Doctrine Table annotation when generating table name

### ISSUE THREAD

**Issue by tom-combet** [PrestaShop/PrestaShop#26810]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

I'm developing a module with Doctrine. I'd like to set a custom table name to avoid conflicts with native or other modules tables.
So I want to be able to add my own prefix in addition to the default database prefix, something like this :
`ps_myprefix_myentity` instead of `ps_myentity` (where `ps_` being the default database prefix).

It is currently impossible because :
- Table name is automatically generated with db prefix and entity class name.
- Database prefix is not added to Doctrine Table annotation value if used and Doctrine can't parse constant + string concatenation in annotation, so I can't do:
`@Table(name={_DB_PREFIX_}myprefix_myentity)`

The only workaround I found is to set a constant in module core class and to call it like this:
`@Table(name=MyModule::MY_ENTITY_TABLE_NAME)`
But it's not very clean.

### Describe the solution you'd like

In my opinion, the table annotation should be considered in the table name generation.
So it would use it instead of entity class name if it is found. I could therefore do:
`@Table(name="myprefix_myentity")` which would generate `ps_myprefix_myentity` (where `ps_` being the default database prefix).

**Comment by hibatallahAouadni** [comment]
Hello @Kaudaj 

Thank you for your suggestion. The Product Team will take it into consideration for future developments. 
Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. 

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 

Thank you

**Comment by matks** [comment]
> Table name is automatically generated with db prefix and entity class name.

Can you explain how do you generate entities? Do you use Doctrine CLI?

**Comment by tom-combet** [comment]
@matks Yes, I'm using [MakerBundle](https://symfony.com/doc/current/doctrine.html#creating-an-entity-class) to generate entities. But it won't change [the way PrestaShop generate table name](https://devdocs.prestashop.com/1.7/modules/concepts/doctrine/#creating-the-database), will it?

**Comment by ytilotti** [comment]
In the documentation of Doctrine there is this information:

> The naming strategy is always overridden by entity mapping such as the `Table` annotation. (https://www.doctrine-project.org/projects/doctrine-orm/en/2.10/reference/namingstrategy.html)

Maybe change implementation in core to use https://www.doctrine-project.org/projects/doctrine-orm/en/2.11/cookbook/sql-table-prefixes.html and not NamingStrategy

**Comment by boo-code** [comment]
@ytilotti's reading is the right one, and the reason is structural rather than an oversight.

`PrestaShopBundle\Service\Database\DoctrineNamingStrategy::classToTableName($className)` receives only a class name. A naming strategy has no access to the mapping, so it cannot see `@Table` at all - and Doctrine consults the strategy only when the mapping does not name the table. That is why the two are mutually exclusive today: name the table and you lose the prefix, keep the prefix and you cannot name the table.

Core has already met this exact problem once, on the other side of the same class. `joinTableName()` reads a `ps_table` key out of the `@JoinTable` options through `getPrestashopTable()` and prepends the prefix to it, precisely so a join table can have a custom name and still be prefixed. That escape hatch is used by zero core entities, so the concept exists in the tree and only the entity-table half is missing.

The cookbook subscriber you linked is the canonical way to get both, because `loadClassMetadata` runs after the mapping is resolved and sees the final table name whether it came from `@Table` or from the strategy. Moving the prefix there and reducing the strategy back to plain `UnderscoreNamingStrategy` would make `@Table(name="myprefix_myentity")` produce `ps_myprefix_myentity`.

Two things to weigh on timing. Core is safe: none of the 24 entities under `src/PrestaShopBundle/Entity/` declares a table name, so no core table would move. Third-party modules are not: any module that writes the prefix itself inside `@Table(name="ps_something")` - which is the only way to get a prefixed custom name today - would become `ps_ps_something`. That makes it a major-version change, or one that has to be opt-in.
